cni/pkg/nodeagent/net.go

```diff
@@ -0,0 +1,381 @@
+		log.Errorf("failed to delete inpod rules %v", err)
+		return fmt.Errorf("failed to delete inpod rules %w", err)
+	}
+
+	if err := removePodFromHostNSIpset(pod, &s.hostsideProbeIPSet); err != nil {
+		log.Errorf("failed to remove pod %s from host ipset, error was: %v", pod.Name, err)
+		return err
+	}
+
+	log.Debug("in pod mode - removing pod from ztunnel")
+	if err := s.ztunnelServer.PodDeleted(ctx, string(pod.UID)); err != nil {
+		log.Errorf("failed to delete pod from ztunnel: %v", err)
+	}
+	return nil
+}
+
+// Delete pod from mesh: pod is deleted. iptables rules will die with it, we just need to update ztunnel
+func (s *NetServer) DelPodFromMesh(ctx context.Context, pod *corev1.Pod) error {
+	log := log.WithLabels("ns", pod.Namespace, "name", pod.Name)
+	log.Debug("Pod is now stopped or opt out... cleaning up.")
+
+	if err := removePodFromHostNSIpset(pod, &s.hostsideProbeIPSet); err != nil {
+		log.Errorf("failed to remove pod %s from host ipset, error was: %v", pod.Name, err)
+		return err
+	}
+
+	log.Info("in pod mode - deleting pod from ztunnel")
+
+	// pod is deleted, clean-up its open netns
+	openNetns := s.currentPodSnapshot.Take(string(pod.UID))
+	if openNetns == nil {
+		log.Warn("failed to find pod netns")
+	}
+
+	if err := s.ztunnelServer.PodDeleted(ctx, string(pod.UID)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *NetServer) syncHostIPSets(ambientPods []*corev1.Pod) error {
+	var addedIPSnapshot []netip.Addr
+
+	for _, pod := range ambientPods {
+		podIPs := util.GetPodIPsIfPresent(pod)
+		if len(podIPs) == 0 {
+			log.Warnf("pod %s does not appear to have any assigned IPs, not syncing with ipset", pod.Name)
+		} else {
+			err := addPodToHostNSIpset(pod, podIPs, &s.hostsideProbeIPSet)
+			if err != nil {
+				return err
+			}
+			addedIPSnapshot = append(addedIPSnapshot, podIPs...)
+		}
+
+	}
+	return pruneHostIPset(sets.New(addedIPSnapshot...), &s.hostsideProbeIPSet)
+}
+
+// addPodToHostNSIpset:
+// 1. get pod manifest
+// 2. look for probes of the 3 kinds
+// 2. Get all pod ips (might be several, v6/v4)
+// 3. update ipsets accordingly
+func addPodToHostNSIpset(pod *corev1.Pod, podIPs []netip.Addr, hostsideProbeSet *ipset.IPSet) error {
+	// Add the pod UID as an ipset entry comment, so we can (more) easily find and delete
+	// all relevant entries for a pod later.
+	podUID := string(pod.ObjectMeta.UID)
+	ipProto := uint8(unix.IPPROTO_TCP)
+
+	var ipsetAddrErrs []error
+
+	// For each pod IP
+	for _, pip := range podIPs {
+		// Add to host ipset
+		log.Debugf("adding pod %s probe to ipset %s with ip %s", pod.Name, hostsideProbeSet.Name, pip)
+		// Add IP/port combo to set. Note that we set Replace to true - a pod ip/port combo already being
+		// in the set is perfectly fine, and something we can always safely overwrite, so we will.
+		if err := hostsideProbeSet.AddIP(pip, ipProto, podUID, true); err != nil {
```

Comment:
What if a pod within same node using hostnetwork, amd the pod access a server pod. Does that work like kubelet probe?

Reply:
Yes, that's a good point that we had discussed previously. As with the previous ambient redirection impl, this solution will not work with hostnetwork pods at all.

It's a downside that `inpod` does not work with hostnetwork pods at all, but on the other side of that, none of Istio's existing or previous solutions really would, either. 

Host network pods are both pretty rare and should not be in the mesh, by definition, so the fact that this doesn't work with them is not a big concern.

Reply:
looks good, so we do not need to manage that kind of traffic at all